### PR TITLE
Implementation of $bvModal

### PR DESCRIPTION
**Pull request by parafeu**
Related to #120 

Hi ! This is my first contribution of an open source project, I hope it goes well!
I've made this implementation using mitt.js as global event bus. 
This could probably be replaced by a native solution, but this library is realy tiny and easy to use.

### CODE REVIEW

File: src/components/BModal.vue [hunk 1]
```diff
@@ -34,9 +32,8 @@
                 :size="buttonSize"
                 :variant="cancelVariant"
                 @click="$emit('cancel')"
+                >{{ cancelTitle }}</b-button
```
Comment (mariusa) [inline src/components/BModal.vue:35]: pls fix formatting: ```</b-button``` and ```>``` should be on the same tag, not separate lines

File: src/components/BModal.vue [hunk 2]
```diff
@@ -45,9 +42,8 @@
                 :size="buttonSize"
                 :variant="okVariant"
                 @click="$emit('ok')"
+                >{{ okTitle }}</b-button
```
Comment (mariusa) [inline src/components/BModal.vue:45]: pls fix formatting: ```</b-button``` and ```>``` should be on the same tag, not separate lines

File: src/components/BModal.vue (hunk 1 above)
Comment (mariusa) [reply src/components/BModal.vue:35]: Did anything change here? If not, please make it same as previous code, so there's no diff.

**Comment by xvaara** [comment]
We could just use a reactive object and watch that, no mitt needed.

But, what is the function of this feature? Or is it just for bootstrapVue compatibility? The why would you need to use a global function instead of the provided ways to show modals?

**Comment by xvaara** [comment]
If you need to use bus, I think it would be better to use https://vueuse.org/core/useeventbus/ as I think we'll use that library at some point.

and @parafeu  you should turn the listener off on unmount, so that they aren't left behind.

**Comment by parafeu** [comment]
> We could just use a reactive object and watch that, no mitt needed.
> 
> But, what is the function of this feature? Or is it just for bootstrapVue compatibility? And why would you need to use a global function instead of the provided ways to show modals?

The purpose of my pull request was to cover issue #120 . The advantage of a global function is that it is available in templates no matter which API is used (option or composition). If needed I can make an alternative with a composable. 



> If you need to use bus, I think it would be better to use https://vueuse.org/core/useeventbus/ as I think we'll use that library at some point.
> 
> and @parafeu you should turn the listener off on unmount, so that they aren't left behind.

Totally forgot the basics here, my bad. I'll make a new pull request shortly using vueuse instead of mitt.

**Comment by mariusa** [comment]
A note on `globalProperties`, quoting Evan You:

> [globalProperties] … is essentially a legacy escape hatch and not for setup function. You should just import things directly or use provide/inject in setup functions.

But I'm not sure how to use `provide/inject` in a library. In my app, `app.vue` has `provide()`, all other `.vue` have `inject()` at top. Here `app.vue` is used for demo, being a library, not an actual app.

**Comment by parafeu** [comment]
From Vue3 plugin documentation
> There is no strictly defined scope for a plugin, but common scenarios where plugins are useful include:
> - Make a resource [injectable](https://vuejs.org/guide/components/provide-inject.html) throughout the app by calling [app.provide()](https://vuejs.org/api/application.html#app-provide).

**Comment by xvaara** [comment]
I'm not a fan of libraries adding to the globalProperties. You can implement it in your own code if you need.

Again, I'm not sure why you can't just use a state for showing modals? Why is this feature even needed? This really fells like it's only for legacy versions of code from vue1 days.

**Comment by parafeu** [comment]
I implemented this feature because it was in the roadmap for 1.0.0. It seems indeed that it is mainly for backward compatibility with Bootstrap Vue 2

**Comment by mariusa** [comment]
In this case, it seems best we revert the PR. Do you agree?
Sorry I merged without investigating more.

Marius

**Comment by parafeu** [comment]
As you wish, I dont mind.

But maybe we should discuss about an alternative (using composable ?), it might be useful to control modal visibility without having access to its state.

**Comment by mariusa** [comment]
> it might be useful to control modal visibility without having access to its state.

Would you please start a Discussion? And detail the use case / need from an app developer. Thanks!

**Comment by Micene09** [comment]
In my company we are using a lot the old `$bvModal`, just because it was the only alternative to have a functional modal, that can be launched from any point of the code-base, not just from a component/view.

Personally, I think that at least we should have a composable alternative, this way we can bring almost the same functionality but with the new "Vue 3 fashion".

I agree with @xvaara about globalProperties, but maybe the whole [project](https://github.com/cdmoro/bootstrap-vue-3/projects/1) should be defined better with more details, even just to help the (growing) community here...I would help but I don't know how.
Do you have a Discord channel? How can we start a discussion on how to improve issue tracking and roadmap?

**Comment by mariusa** [comment]
How about having a modal instance in app.vue, and using an event bus (mitt)
to emit hide/show events from anywhere?
app.vue listens to these and hides/shows with the received content .

For discussions, there's the Discussions section here on Github. Spreading
even thinner on other apps won't benefit.

**Comment by Micene09** [comment]
> How about having a modal instance in app.vue, and using an event bus (mitt)
to emit hide/show events from anywhere?
app.vue listens to these and hides/shows with the received content .

Not atomic at all...
I'm taking about a "composable", spawning a modal from anywhere.
Imagine the equivalent of the $bvModal, returning a Promise that gets resolved once the modal was closed.
In so many business logics, outside a view and components logics, during a very complex Store (async) Action may happen to give a feedback using an info/warning (or whatever) modal with a message.

>For discussions, there's the Discussions section here on Github. Spreading
even thinner on other apps won't benefit.

Sorry, my bad.
I'm ready to start a new discussion or just move this thread the one creates by @parafeu.

**Pull request merged**